### DISCUSSION TITLE
Creating custom math functions

### DISCUSSION THREAD

**Discussion by patrick-kidger** [typst/typst#997] [General]
I'd like to write a derivative:
```typst
$ (upright(d) x)/(upright(d) t) $ 
```
and to simplify things a little I'm defining a custom function like so:
```typst
#let d(value) = [$upright(d)#value$]

$ #d[x]/#d[t] $
```

However I think each `#d` function call implies leaving math mode only to re-enter it again, much like doing `$ ... \text{$ ... $} ... $` in LaTeX. So I just wanted to check if this was really the best way to do things?

**Comment by PgBiel** [comment]
Don't worry, your function will work just like any other function in math mode (no need to switch to code mode). That is, you can write: (EDIT: This example is wrong; see my reply below.)

```
#let d(value) = $upright(d)#value$

$ d(x)/d(t) $
```

(Note that the `[ ]` around the `$ $` were also unnecessary, as `$ $` already generates content.)

Additionally, note that the function you're writing is already in typst, as a symbol (in fact, this didn't have to be a function at all, just the 'd' symbol):

```
$ (dif x) / (dif t) $
```

**Reply by patrick-kidger** [reply]
Thanks for the reply!

To make it work I had to tweak your first example:
```typst
#let d(value) = $dif#value$

$ #d[x]/#d[t] $
```
(the version you've got prints out the brackets without doing a function call)
But I think this is avoiding the math-and-back-against issue I originally had!

This is probably what I'll use going forward, as it's a couple of characters shorter than writing the `(dif ...)` manually.

**Reply by PgBiel** [reply]
Oh, sorry @patrick-kidger . The problem here seems to be that the function has just one letter as its name, so it's being recognized as a variable (and the function itself isn't being called). This seems to work as you'd expect:

```rust
#let dd(value) = $upright(d)#value$

$ dd(x)/dd(t) $
```

In general, use more than one letter as your math function's name.

**Reply by patrick-kidger** [reply]
Ah, marvellous! I didn't know about that footgun.

Thanks for your help, `dd` is what I'll use.

**Reply by PgBiel** [reply]
No problem! Keep in mind that this also applies to code variables (not just functions) in a math context (they will only be inserted when they have more than one letter as a name).

**Comment by Leedehai** [comment]
Hi, a side note: you may try out my [physics](https://github.com/Leedehai/typst-physics) package for derivatives. Hope u find it useful :)

**Comment by YDX-2147483647** [comment]
Here's another trick.

```typst
#let dv(f, x) = $(dif #f) / (dif #x)$

$ dv("", x) f exp g = dv(exp g, x) + dv(f, x) exp g $
```

![](https://user-images.githubusercontent.com/73375426/235844602-7e52fd70-3092-464d-83c6-2a02b1dc7b8c.png)

---

**Update**: The [physics](https://github.com/Leedehai/typst-physics) package (metioned above) has [a more powerful `dv`](https://github.com/Leedehai/typst-physics/blob/37785987bec96672d35289dfefefbb24d16749a9/physics.typ#L502-L533).
![physics](https://user-images.githubusercontent.com/73375426/235850187-f9266a71-603b-4f14-98c5-c43cfb289ab2.png)
